Update ads

tiagopereira-sage opened the pull request.

tiagoafp committed:
Merge remote-tracking branch 'kaltura/develop' into develop

tiagoafp committed:
Update youboraLib and add new ads events

tiagoafp committed:
remove logs

OrenMe commented on Sources/PKYouboraAdsAdapter.swift:
```diff
@@ -156,11 +186,19 @@ extension PKYouboraAdsAdapter {
+                    
+                    if let adapter = self.plugin?.adapter, adapter.flags.started {
+                        self.fireStart()
+                    }
```
all of this is basically incorrect in this event context.
Ad loaded may happen at any given time before ad playback due to load optimisations.
Actual ad start event is signaled by the `adStarted` event(just below this).
The only thing which is relevant here is the adInfo initialization.
The start of ad is `adStarted` and the join is the first `adDidProgressToTime` event - just like we did in the JS plugin.

OrenMe replied:
also - what is the diff between `fireStart` and `tryAdBreakStart` method calls?

tiagopereira-sage replied:
Just changed this logic to `adStarted`

OrenMe commented on Sources/PKYouboraAdsAdapter.swift:
```diff
@@ -171,17 +209,15 @@ extension PKYouboraAdsAdapter {
-                    // Has no effect when already playing ad and resumed because ad was already started.
-                    self.plugin?.adapter?.fireStart()
-                    self.fireStart()
                     self.fireJoin()
```
why do we need the join here?

tiagopereira-sage replied:
After review it and since the `fireJoin` was moved to `adDidProgressToTime` it makes no sense to be here so I changed this as well

OrenMe commented on Sources/PKYouboraAdsAdapter.swift:
```diff
@@ -214,18 +251,19 @@ extension PKYouboraAdsAdapter {
                     self.lastReportedResource = event.adTagUrl
                 }
-            // When ad request the content to resume (finished or error)
+                // When ad request the content to resume (finished or error)
```
fix indentation

OrenMe commented on Sources/PKYouboraAdsAdapter.swift:
```diff
@@ -214,18 +251,19 @@ extension PKYouboraAdsAdapter {
                 messageBus.addObserver(self, events: [e.self]) { [weak self] event in
                     guard let self = self else { return }
-
+                    
```
remove empty line

OrenMe commented on Sources/PKYouboraAdsAdapter.swift:
```diff
@@ -214,18 +251,19 @@ extension PKYouboraAdsAdapter {
             case let e where e.self == AdEvent.adDidRequestContentResume:
                 messageBus.addObserver(self, events: [e.self]) { [weak self] event in
                     guard let self = self else { return }
                     self.fireStop()
+                    self.tryAdBreakStop()
```
what is the purpose of `adBreakStop`  with `firstStop`.
Also, why do we have `fireStop` here? we should have it in completed or error cases already.

OrenMe commented on Sources/PKYouboraAdsAdapter.swift:
```diff
@@ -240,6 +278,36 @@ extension PKYouboraAdsAdapter {
+                    
+                    self.fireQuartile(3)
+                }
+            case let e where e.self == AdEvent.adBreakStarted:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    guard let self = self else { return }
+                    
+                    self.fireAdBreakStart()
+                }
+            case let e where e.self == AdEvent.adBreakEnded:
+                messageBus.addObserver(self, events: [e.self]) { [weak self] event in
+                    guard let self = self else { return }
+                    
+                    self.fireAdBreakStop()
```
we only fire these events here - I think the other places with the try logic are redundant.
The event model should be predictable.

tiagopereira-sage replied:
Maybe is easier to try to explain here why the `adBreaks` are not only here and we have the `tryAdBreak` methods as well. This `AdEvent.adBreakStarted` and `AdEvent.adBreakEnded` are only available for IMA DAI. But for instance in IMA you can have as well an ad break with more then 1 pod and we consider that an ad break as well. So the workaround that we found to catch the adBreaks in IMA is check the pods and fire the `adBreakStart` when the first ad of the pod will be started and the `fireAdBreakStop` when the last pod of this ad interval ends. That's the reason why every time an ad starts or ends we check if if the first or the last pod and we trigger the event. This way we can have adBreaks working with IMA and IMA DAI.

OrenMe commented on Sources/PKYouboraAdsAdapter.swift:
```diff
@@ -198,6 +234,7 @@ extension PKYouboraAdsAdapter {
                 messageBus.addObserver(self, events: [e.self]) { [weak self] event in
                     guard let self = self else { return }
                     self.fireStop(["skipped":"true"])
+                    self.tryAdBreakStop()
```
skip should result in adBreakEnded, why do we need to have try logic all around?

tiagopereira-sage replied:
The explanation above

OrenMe reviewed (changes requested):
There are some new features missing:
* getGivenBreaks
* getExpectedPattern
* getBreaksTime
* getCampaign
* getCreativeId
* getProvider
* fireManifest event

Are we planing to add them? 
You can work with @x-NR-x to map the corresponding events to get the data points.

OrenMe commented on Sources/PKYouboraAdsAdapter.swift:
```diff
@@ -198,6 +234,7 @@ extension PKYouboraAdsAdapter {
                 messageBus.addObserver(self, events: [e.self]) { [weak self] event in
                     guard let self = self else { return }
                     self.fireStop(["skipped":"true"])
```
I know JS changed this to `fireSkip`, is this also same for iOS?

tiagopereira-sage replied:
you are right, just changed this for `fireSkip`

tiagoafp committed:
Improve code from the PR

Add fireSkip, remove fireJoin from resume and add it to ptogressToTime, Clean all the logic from adLoaded, improve indentation

tiagoafp committed:
Add some extra ad options

tiagoafp committed:
Track cue points to check number of ad breaks and start time

tiagoafp committed:
Improve adBreaks and add getter to ad provider

tiagoafp committed:
Update youboraLib to support fastData injection

tiagoafp committed:
Validate code and host on fastDataConfig

tiagoafp committed:
Improve indentation
